### PR TITLE
[enhance-wip](multi-catalog) Speed up consume rate of hms events.

### PR DESCRIPTION
**Pull request by dutyu**
## Proposed changes

The current implement will persist all catalogs/databases of external catalogs, and only the master FE can handle hms events and make all slave nodes replay these events, this will bring some problems:

- The hms event processor ( `MetastoreEventsProcessor` ) can not consume the events in time. (Add journal log is a synchronized method, we can not speed up the consume rate by using concurrent processing, and each add-journal-log operation costs about tens of milliseconds) So the meta info of hive maybe out of date.

- Slave FE nodes maybe crashed if FE replays the journal logs of hms events failed. (In fact we have fixed some issues about this, but we can not make sure all the issues have been resolved)

- There are many journal logs which are produced by hms events, but in fact these logs are not used anymore after FE restart. It makes the start time of all FE nodes very long.

Now doris try to persis all databases/tables of external catalogs just to make sure that the dbId/tableId of databases/tables are the same through all FE nodes, it will be used by analysis jobs. 

In this pr, we use a meta id manager called `ExternalMetaIdMgr` to manage these meta ids. On every loop when master fetches a batch of hms events, it handles the meta ids first and produce only one meta id mappings log, slave FE nodes will replay this log to sync the changes about these meta ids. `MetastoreEventsProcessor` will start on every FE nodes and try to consume these hms events as soon as possible.

## Further comments

I've submitted two prs ( #22869 #21589 ) to speed up the consume rate of hms events before, it works fine when there are many `AlterTableEvent` / `DropTableEvent` on hive cluster. But the improvement is not that significant when most of hms events are partition-events.  Unfortunately, we performed a cluster upgrade (upgrade spark 2.x to spark 3.x), maybe this is the reason that resulting in the majority of Hive Metastore events became partition-events. This is also the reason for the existence of this pull request.

Based on our observation, after merging this pull request, Doris is now capable of processing thousands of Hive Metastore events per second, compared to the previous capability of handling only a few dozen events.

```java
2023-12-07 05:17:03,518 INFO (replayer|105) [Env.replayJournal():2614] replayed journal id is 18287902, replay to journal id is 18287903
2023-12-07 05:17:03,735 INFO (org.apache.doris.datasource.hive.event.MetastoreEventsProcessor|37) [MetastoreEventFactory.mergeEvents():188] Event size on catalog [xxx] before merge is [1947], after merge is [1849]
2023-12-07 05:17:03,735 INFO (org.apache.doris.datasource.hive.event.MetastoreEventsProcessor|37) [MetastoreEvent.infoLog():193] EventId: 357955309 EventType: ALTER_PARTITION catalogName:[xxx],dbName:[xxx],tableName:[xxx],partitionNameBefore:[partitions=2022-05-27],partitionNameAfter:[partitions=2022-05-27]
2023-12-07 05:17:03,735 INFO (org.apache.doris.datasource.hive.event.MetastoreEventsProcessor|37) [MetastoreEvent.infoLog():193] EventId: 357955310 EventType: ALTER_PARTITION catalogName:[xxx],dbName:[xxx],tableName:[xxx],partitionNameBefore:[pday=20230318],partitionNameAfter:[pday=20230318]
2023-12-07 05:17:03,735 INFO (org.apache.doris.datasource.hive.event.MetastoreEventsProcessor|37) [MetastoreEvent.infoLog():193] EventId: 357955311 EventType: ALTER_PARTITION catalogName:[xxx],dbName:[xxx],tableName:[xxx],partitionNameBefore:[pday=20190826],partitionNameAfter:[pday=20190826]
2023-12-07 05:17:03,735 INFO (org.apache.doris.datasource.hive.event.MetastoreEventsProcessor|37) [MetastoreEvent.infoLog():193] EventId: 357955312 EventType: ALTER_PARTITION catalogName:[xxx],dbName:[xxx],tableName:[xxx],partitionNameBefore:[partitions=2021-09-16],partitionNameAfter:[partitions=2021-09-16]
2023-12-07 05:17:03,735 INFO (org.apache.doris.datasource.hive.event.MetastoreEventsProcessor|37) [MetastoreEvent.infoLog():193] EventId: 357955314 EventType: ALTER_PARTITION catalogName:[xxx],dbName:[xxx],tableName:[xxx],partitionNameBefore:[partitions=2020-04-26],partitionNameAfter:[partitions=2020-04-26]
2023-12-07 05:17:03,735 INFO (org.apache.doris.datasource.hive.event.MetastoreEventsProcessor|37) [MetastoreEvent.infoLog():193] EventId: 357955315 EventType: ALTER_PARTITION catalogName:[xxx],dbName:[xxx],tableName:[xxx],partitionNameBefore:[pday=20230702],partitionNameAfter:[pday=20230702]
2023-12-07 05:17:03,735 INFO (org.apache.doris.datasource.hive.event.MetastoreEventsProcessor|37) [MetastoreEvent.infoLog():193] EventId: 357955317 EventType: ALTER_PARTITION catalogName:[xxx],dbName:[xxx],tableName:[xxx],partitionNameBefore:[pday=20211019],partitionNameAfter:[pday=20211019]
...
2023-12-07 05:17:03,989 INFO (org.apache.doris.datasource.hive.event.MetastoreEventsProcessor|37) [MetastoreEvent.infoLog():193] EventId: 357957252 EventType: ALTER_PARTITION catalogName:[xxx],dbName:[xxx],tableName:[xxx],partitionNameBefore:[partitions=2021-08-27],partitionNameAfter:[partitions=2021-08-27]
2023-12-07 05:17:03,989 INFO (org.apache.doris.datasource.hive.event.MetastoreEventsProcessor|37) [MetastoreEvent.infoLog():193] EventId: 357957253 EventType: ALTER_PARTITION catalogName:[xxx],dbName:[xxx],tableName:[xxx],partitionNameBefore:[partitions=2022-02-05],partitionNameAfter:[partitions=2022-02-05]
2023-12-07 05:17:04,661 INFO (replayer|105) [Env.replayJournal():2614] replayed journal id is 18287903, replay to journal id is 18287904
2023-12-07 05:17:05,028 INFO (org.apache.doris.datasource.hive.event.MetastoreEventsProcessor|37) [MetastoreEventsProcessor.realRun():116] Events size are 587 on catalog [xxx]
2023-12-07 05:17:05,662 INFO (org.apache.doris.datasource.hive.event.MetastoreEventsProcessor|37) [MetastoreEventFactory.mergeEvents():188] Event size on catalog [xxx] before merge is [587], after merge is [587]
```

### CODE REVIEW

**Comment by dutyu** [comment]
run buildall

**Comment by dutyu** [comment]
run buildall

**Comment by dutyu** [comment]
run buildall

**Comment by dutyu** [comment]
run buildall

**Comment by dutyu** [comment]
run buildall

**Comment by dutyu** [comment]
run buildall

**Comment by dutyu** [comment]
run buildall

**Comment by dutyu** [comment]
run feut

**Comment by dutyu** [comment]
run buildall

**Comment by dutyu** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/datasource/ExternalMetaIdMgr.java [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+
+    public @Nullable DbMetaIdMgr getDbMetaIdMgr(long catalogId, String dbName) {
+        CtlMetaIdMgr ctlMetaIdMgr = idToCtlMgr.get(catalogId);
+        if (ctlMetaIdMgr == null) {
+            return null;
+        }
+        return ctlMetaIdMgr.dbNameToMgr.get(dbName);
+    }
+
+    public @Nullable TblMetaIdMgr getTblMetaIdMgr(long catalogId, String dbName, String tblName) {
```
[the remark this answers is not in the document]
Comment (dutyu) [reply fe/fe-core/src/main/java/org/apache/doris/datasource/ExternalMetaIdMgr.java]: done

**Comment by dutyu** [comment]
run buildall

**Comment by dutyu** [comment]
run buildall

**Comment by dutyu** [comment]
run buildall

**Comment by dutyu** [comment]
run buildall

**Comment by dutyu** [comment]
run feut

**Comment by dutyu** [comment]
run buildall

**Comment by dutyu** [comment]
run buildall

**Comment by dutyu** [comment]
run buildall

**Comment by dutyu** [comment]
run feut

**Comment by morningman** [comment]
run buildall

**Comment by morningman** [comment]
run buildall

**Comment by dutyu** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/datasource/ExternalMetaIdMgr.java [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+ * Now it just manages the external meta ids of hms events,
+ * but it will be extended to manage other external meta ids in the future.
+ * </pre>
+ * TODO: remove InitCatalogLog and InitDatabaseLog, manage external meta ids at ExternalMetaIdMgr
+ */
+public class ExternalMetaIdMgr {
+
+    private static final Logger LOG = LogManager.getLogger(ExternalMetaIdMgr.class);
+
+    private final Map<Long, CtlMetaIdMgr> idToCtlMgr = Maps.newConcurrentMap();
```
Comment (zddr) [inline fe/fe-core/src/main/java/org/apache/doris/datasource/ExternalMetaIdMgr.java:44]: not persist?

File: fe/fe-core/src/main/java/org/apache/doris/datasource/MetaIdMappingsLog.java [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+import java.util.Objects;
+
+@Data
+public class MetaIdMappingsLog implements Writable {
+
+    @SerializedName(value = "catalogId")
+    private long catalogId = -1L;
+
+    @SerializedName(value = "fromHmsEvent")
+    private boolean fromHmsEvent = false;
```
[the remark this answers is not in the document]
Comment (dutyu) [reply fe/fe-core/src/main/java/org/apache/doris/datasource/MetaIdMappingsLog.java:43]: Maybe we can remove all init catalog/database logs and use `MetaIdMappingsLog` for instead in the future.

File: fe/fe-core/src/main/java/org/apache/doris/datasource/MetaIdMappingsLog.java [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+            this.opType = opType;
+            this.metaObjType = metaObjType;
+            this.dbName = dbName;
+            this.tblName = tblName;
+            this.partitionName = partitionName;
+            this.id = id;
+        }
+
+        @Override
+        public void write(DataOutput out) throws IOException {
```
[the remark this answers is not in the document]
Comment (dutyu) [reply fe/fe-core/src/main/java/org/apache/doris/datasource/MetaIdMappingsLog.java:176]: write / read not need means?

File: fe/fe-core/src/main/java/org/apache/doris/datasource/ExternalMetaIdMgr.java (hunk 2 above)
Comment (dutyu) [reply fe/fe-core/src/main/java/org/apache/doris/datasource/ExternalMetaIdMgr.java:44]: We just need to ensure that these meta ids are the same through different FE nodes, so i think we do not need persist here.

File: fe/fe-core/src/main/java/org/apache/doris/datasource/hive/event/MetastoreEventFactory.java [hunk 5]
```diff
@@ -77,23 +79,67 @@ List<MetastoreEvent> getMetastoreEvents(List<NotificationEvent> events, HMSExter
...
+    }
+
+    private void logMetaIdMappings(long catalogId, long lastSyncedEventId, List<MetastoreEvent> mergedEvents) {
+        MetaIdMappingsLog log = new MetaIdMappingsLog();
+        log.setCatalogId(catalogId);
+        log.setFromHmsEvent(true);
+        log.setLastSyncedEventId(lastSyncedEventId);
+        for (MetastoreEvent event : mergedEvents) {
+            switch (event.getEventType()) {
+                case CREATE_DATABASE:
```
[the remark this answers is not in the document]
Comment (dutyu) [reply fe/fe-core/src/main/java/org/apache/doris/datasource/hive/event/MetastoreEventFactory.java:96]: When there are some new `create db/create table/create partition` events being produced, master `FE` assigns an id for this database/table/partition and send it to slave `FE` nodes through `MetaIdMappingsLog`, so master and slave `FE` nodes can keep consistent meta ids (dbId/tblId).

**Comment by dutyu** [comment]
run buildall

**Comment by dutyu** [comment]
run buildall

**Comment by dutyu** [comment]
run buildall

**Comment by dutyu** [comment]
run buildall

**Comment by dutyu** [comment]
run beut

**Comment by dutyu** [comment]
run buildall

**Review by zddr — APPROVED** [review]

**Comment by dutyu** [comment]
run buildall

**Comment by dutyu** [comment]
run buildall

**Review by zddr — APPROVED** [review]

**Comment by dutyu** [comment]
run compile

**Comment by dutyu** [comment]
run feut

**Comment by dutyu** [comment]
run feut

**Comment by dutyu** [comment]
run buildall

**Review by morningman — DISMISSED** [review]
LGTM

**Comment by morningman** [comment]
run buildall

**Comment by morningman** [comment]
run buildall

**Review by morningman — APPROVED** [review]
LGTM

**Pull request merged**